xds/src/main/java/io/grpc/xds/CsdsService.java

```diff
@@ -150,124 +141,33 @@ private ClientStatusResponse getConfigDumpForRequest(ClientStatusRequest request
-      Map<String, ResourceMetadata> resourcesMetadata, String version) {
-    ClustersConfigDump.Builder cdsConfig = ClustersConfigDump.newBuilder();
-    for (ResourceMetadata metadata : resourcesMetadata.values()) {
-      cdsConfig.addDynamicActiveClusters(buildDynamicCluster(metadata));
-    }
-    return cdsConfig.setVersionInfo(version).build();
-  }
-
-  @VisibleForTesting
-  static DynamicCluster buildDynamicCluster(ResourceMetadata metadata) {
-    DynamicCluster.Builder dynamicCluster = DynamicCluster.newBuilder()
-        .setVersionInfo(metadata.getVersion())
-        .setClientStatus(metadataStatusToClientStatus(metadata.getStatus()))
-        .setLastUpdated(Timestamps.fromNanos(metadata.getUpdateTimeNanos()));
-    if (metadata.getErrorState() != null) {
-      dynamicCluster.setErrorState(metadataUpdateFailureStateToProto(metadata.getErrorState()));
-    }
-    if (metadata.getRawResource() != null) {
-      dynamicCluster.setCluster(metadata.getRawResource());
-    }
-    return dynamicCluster.build();
-  }
-
-  @VisibleForTesting
-  static EndpointsConfigDump dumpEdsConfig(Map<String, ResourceMetadata> resourcesMetadata) {
-    EndpointsConfigDump.Builder edsConfig = EndpointsConfigDump.newBuilder();
-    for (ResourceMetadata metadata : resourcesMetadata.values()) {
-      edsConfig.addDynamicEndpointConfigs(buildDynamicEndpointConfig(metadata));
-    }
-    return edsConfig.build();
-  }
-
-  @VisibleForTesting
-  static DynamicEndpointConfig buildDynamicEndpointConfig(ResourceMetadata metadata) {
-    DynamicEndpointConfig.Builder dynamicRouteConfig = DynamicEndpointConfig.newBuilder()
-        .setVersionInfo(metadata.getVersion())
-        .setClientStatus(metadataStatusToClientStatus(metadata.getStatus()))
-        .setLastUpdated(Timestamps.fromNanos(metadata.getUpdateTimeNanos()));
-    if (metadata.getErrorState() != null) {
-      dynamicRouteConfig.setErrorState(metadataUpdateFailureStateToProto(metadata.getErrorState()));
-    }
-    if (metadata.getRawResource() != null) {
-      dynamicRouteConfig.setEndpointConfig(metadata.getRawResource());
+    ClientConfig.Builder builder = ClientConfig.newBuilder()
+        .setNode(xdsClient.getBootstrapInfo().node().toEnvoyProtoNode());
+    for (ResourceType type : ResourceType.values()) {
+      if (type == ResourceType.UNKNOWN) {
+        continue;
+      }
+      Map<String, ResourceMetadata> metadataMap = xdsClient.getSubscribedResourcesMetadata(type);
+      for (String resourceName : metadataMap.keySet()) {
+        ResourceMetadata metadata = metadataMap.get(resourceName);
+        GenericXdsConfig.Builder genericXdsConfigBuilder = GenericXdsConfig.newBuilder()
+            .setTypeUrl(type.typeUrl())
+            .setName(resourceName)
+            .setClientStatus(metadataStatusToClientStatus(metadata.getStatus()));
+        if (metadata.getRawResource() != null) {
+          genericXdsConfigBuilder
+              .setVersionInfo(metadata.getVersion())
```

Comment:
@ejona86 Unless I'm missing something, there's no issue with `getCurrentVersion()` anymore, because there's no `getCurrentVersion()` :) Ref https://github.com/grpc/grpc-java/issues/8642

Reply:
Yeah, looks like that part of #8642 is resolved. But the lack of thread safety of `getSubscribedResourcesMetadata()` is still present.